README.md
<!-- Improved compatibility of back to top link: See: https://github.com/othneildrew/Best-README-Template/pull/73 -->

<a name="readme-top"></a>

<!--
*** Thanks for checking out the Best-README-Template. If you have a suggestion
*** that would make this better, please fork the repo and create a pull request
*** or simply open an issue with the tag "enhancement".
*** Don't forget to give the project a star!
*** Thanks again! Now go create something AMAZING! :D
-->

<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/veraposeidon/SubtitleEasy">
    <img src="https://res.cloudinary.com/dgchmgebr/image/upload/v1688307543/logo-big_g40trq.png" alt="Logo" width="160" height="160">
  </a>

<h3 align="center">SubtitleEasy</h3>

  <p align="center">
    This is a Chrome extension designed to provide bilingual subtitles for videos, aiming to enhance your viewing and learning experience in different language environments.
    <br />
    这是一个Chrome扩展程序，旨在为视频提供双语字幕功能，以帮助您在不同语言环境下观看和学习时获得更出色的体验。
    <br />
    これは、ビデオに対してバイリンガルの字幕機能を提供するためのChrome拡張機能です。異なる言語環境での視聴や学習の体験を向上させることを目的としています。
    <br />
    <a href="https://github.com/veraposeidon/SubtitleEasy"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/veraposeidon/SubtitleEasy">View Demo</a>
    ·
    <a href="https://github.com/veraposeidon/SubtitleEasy/issues">Report Bug</a>
    ·
    <a href="https://github.com/veraposeidon/SubtitleEasy/issues">Request Feature</a>
  </p>
</div>

<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#how-to-download-and-install-the-chrome-extension">How to Download and Install the Chrome Extension</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#step-1-download-the-zip-file">Step 1: Download the Zip File</a></li>
        <li><a href="#step-2-extract-the-zip-file">Step 2: Extract the Zip File</a></li>
        <li><a href="#step-3-install-the-chrome-extension">Step 3: Install the Chrome Extension</a></li>
        <li><a href="#step-4-enable-the-chrome-extension">Step 4: Enable the Chrome Extension</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->

## About The Project

Current supported video platforms:

- [Apple Developer Videos (WWDC sessions)](https://developer.apple.com/videos/)
- [deeplearning.ai](https://www.deeplearning.ai/)

[![Product Name Screen Shot][product-screenshot]](https://github.com/veraposeidon/SubtitleEasy)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

### Built With

- Javascript
- [Chrome Extension Manifest V3](https://developer.chrome.com/docs/extensions/mv3/)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->

## How to Download and Install the Chrome Extension

This tutorial will guide you through the process of downloading and installing the Chrome extension from the provided zip file.

### Prerequisites

Google Chrome browser installed on your computer.

### Step 1: Download the Zip File

1. Go to the [Releases](https://github.com/veraposeidon/SubtitleEasy/releases) page.
2. Find the latest release version and click on the "SubtitleEasy.zip" link to download the zip file.
3. Save the zip file to a location on your computer.

### Step 2: Extract the Zip File

1. Locate the downloaded zip file on your computer.
2. Right-click on the zip file and select "Extract All" or a similar option.
3. Choose a destination folder where you want to extract the contents of the zip file.
4. Click "Extract" to extract the files from the zip archive.

### Step 3: Install the Chrome Extension

1. Open the Google Chrome browser on your computer.
2. In the address bar, type chrome://extensions and press Enter.
3. In the top right corner, enable "Developer mode".
4. Click on the "Load unpacked" button.
5. Navigate to the folder where you extracted the zip file in Step 2.
6. Select the "SubtitleEasy" folder and click "Select Folder".
7. The Chrome extension will now be installed and ready to use.

### Step 4: Enable the Chrome Extension

1. After installation, you should see the Chrome extension's icon appear in the browser's toolbar. If you don't see the icon, click on the puzzle piece icon in the toolbar and pin the extension.
2. Click on the Chrome extension's icon to open the popup window.
3. Click on the "Enable" button to enable the Chrome extension.
4. Select the language you want to use for the subtitles.
5. Now you can go to the Apple Developer website or Coursera and watch videos with subtitles.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->

## Usage

Use this space to show useful examples of how a project can be used. Additional screenshots, code examples and demos work well in this space. You may also link to more resources.

<img src="https://res.cloudinary.com/dgchmgebr/image/upload/v1688307470/usage-1_lfhc1m.png" alt="usage-1" width="800" height="400">

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ROADMAP -->

## Roadmap

- [x] support dual subtitle for WWDC session
- [ ] support i18n for popup page
- [ ] support dual subtitle Coursera videos
- [ ] support dual subtitle Deeplearning.ai videos

See the [open issues](https://github.com/veraposeidon/SubtitleEasy/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTRIBUTING -->

## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- LICENSE -->

## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- CONTACT -->

## Contact

hi.xiaohai - [@veraposeidon](https://twitter.com/veraposeidon) - veraposeidon@gmail.com

Project Link: [https://github.com/veraposeidon/SubtitleEasy](https://github.com/veraposeidon/SubtitleEasy)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- ACKNOWLEDGMENTS -->

## Acknowledgments

- [GoogleChrome/chrome-extensions-samples: Chrome Extensions Samples](https://github.com/GoogleChrome/chrome-extensions-samples)
- [tamshadow/coursera-subtitle-translation](https://github.com/tamshadow/coursera-subtitle-translation)

## Inspiration

- [YouTube™ Dual Caption](https://www.dual-subtitles.com/)
- [Dualsub](https://www.dualsub.xyz/)

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->

[contributors-shield]: https://img.shields.io/github/contributors/veraposeidon/SubtitleEasy.svg?style=for-the-badge
[contributors-url]: https://github.com/veraposeidon/SubtitleEasy/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/veraposeidon/SubtitleEasy.svg?style=for-the-badge
[forks-url]: https://github.com/veraposeidon/SubtitleEasy/network/members
[stars-shield]: https://img.shields.io/github/stars/veraposeidon/SubtitleEasy.svg?style=for-the-badge
[stars-url]: https://github.com/veraposeidon/SubtitleEasy/stargazers
[issues-shield]: https://img.shields.io/github/issues/veraposeidon/SubtitleEasy.svg?style=for-the-badge
[issues-url]: https://github.com/veraposeidon/SubtitleEasy/issues
[license-shield]: https://img.shields.io/github/license/veraposeidon/SubtitleEasy.svg?style=for-the-badge
[license-url]: https://github.com/veraposeidon/SubtitleEasy/blob/master/LICENSE.txt
[product-screenshot]: https://res.cloudinary.com/dgchmgebr/image/upload/v1688307470/usage-2_c2tezp.png


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/veraposeidon/SubtitleEasy/compare/v1.1.0...v1.2.0) (2023-07-05)

### Features

- remove uncomple coursera feature ([5df375f](https://github.com/veraposeidon/SubtitleEasy/commit/5df375f34a29aa0463a6d80e3871b1d270b13d22))
- support deeplearning.ai courses ([eea50f1](https://github.com/veraposeidon/SubtitleEasy/commit/eea50f1b2efa1cfa996374e021c3ab78d79a2a70))

## 1.1.0 (2023-07-02)

### Features

- support wwdc session dual subtitles ([5708d7c](https://github.com/veraposeidon/SubtitleEasy/commit/5708d7cb3da3cf184005f60ebf0ec8d861f2bf84))

### Bug Fixes

- set default target language for first installation ([b472645](https://github.com/veraposeidon/SubtitleEasy/commit/b47264587a7f116899303c11e7aadbb36e76c047))


webpack.config.js
const path = require('path');
const CopyWebpackPlugin = require('copy-webpack-plugin');
const { WebpackManifestPlugin } = require('webpack-manifest-plugin');

module.exports = {
  mode: 'production',
  entry: {
    // No entry point needed if you only want to copy files
  },
  plugins: [
    new WebpackManifestPlugin({
      fileName: 'manifest.json',
      seed: require('./subtitle.dual/manifest.json'),
      generate: (seed) => {
        const manifest = seed || {};

        // set version to current git commit tag and rm the prefix 'v'
        const version = require('child_process')
          .execSync('git describe --tags --always')
          .toString()
          .trim()
          .replace(/^v/, '');

        console.log('version', version);
        manifest.version = version.toString();

        return manifest;
      }
    }),
    new CopyWebpackPlugin({
      // copy subtitle.dual folder to dist except manifest.json
      patterns: [
        {
          from: path.resolve(__dirname, 'subtitle.dual'),
          to: path.resolve(__dirname, 'dist'),
          globOptions: {
            ignore: ['**/manifest.json']
          }
        }
      ]
    })
  ],
  devtool: 'source-map'
};


subtitle.dual/popup.js
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const toggleSwitch = document.getElementById('toggle-switch');

// 读取用户设置，更新 UI
chrome.storage.sync.get('enabled', ({ enabled }) => {
  toggleSwitch.checked = enabled;
  // 更新 UI
  updatePopupStatus(enabled);
});

// 监听开关切换事件，更新用户设置
toggleSwitch.addEventListener('change', async ({ target }) => {
  // 更新 UI
  updatePopupStatus(target.checked);
  // 保存用户设置
  chrome.storage.sync.set({ enabled: target.checked }).then(() => {});
  // 通知 content 更新字幕
  const [tab] = await chrome.tabs.query({ active: true, currentWindow: true });
  // 拦截非声明页面的请求
  if (!tab.url) {
    return;
  }
  // 发消息
  await chrome.tabs.sendMessage(tab.id, {
    enabled: target.checked
  });
});

// 配置下拉框数据
document.addEventListener('DOMContentLoaded', async () => {
  const select = document.getElementById('language-select');
  const options = await getSupportLanguages();
  console.log('options', options);
  Object.keys(options).forEach((key) => {
    const optionElement = document.createElement('option');
    optionElement.value = key;
    optionElement.innerText = options[key];
    select.appendChild(optionElement);
  });

  // 读取用户设置，更新下拉框默认值
  chrome.storage.sync.get('target_language', ({ target_language }) => {
    if (!target_language) {
      const locale = chrome.i18n.getUILanguage();
      target_language = locale;
    }
    select.value = target_language;
  });

  // 监听下拉框切换事件，更新用户设置
  select.addEventListener('change', async ({ target }) => {
    // 保存用户设置
    chrome.storage.sync.set({ target_language: target.value }).then(() => {});
    // 通知 content 更新字幕
    const [tab] = await chrome.tabs.query({
      active: true,
      currentWindow: true
    });
    // 拦截非声明页面的请求
    if (!tab.url) {
      return;
    }
    // toast 提示
    showToast('设置生效，刷新网页');
    // reload tab
    await chrome.tabs.reload(tab.id);
  });
});

function updatePopupStatus(enabled) {
  // chrome.action.setBadgeText({ text: enabled ? 'ON' : 'OFF' }).then(() => {});
  // green and gray
  chrome.action.setBadgeBackgroundColor({
    color: enabled ? '#00ff00' : '#808080'
  });
  // disable select if enabled is false
  const select = document.getElementById('language-select');
  select.disabled = !enabled;
}

async function getSupportLanguages() {
  const recent_locale = (await chrome.storage.sync.get('recent_locale'))
    .recent_locale;
  const current_locale = chrome.i18n.getUILanguage();
  await chrome.storage.sync.set({
    recent_locale: current_locale
  });
  const locale_same = recent_locale === current_locale;
  const support_languages = (await chrome.storage.sync.get('support_languages'))
    .support_languages;
  if (support_languages || locale_same) {
    return support_languages;
  } else {
    return getGoogleTranslateApiSupportedLanguages(current_locale);
  }
}

/// 获取谷歌翻译支持的语言，保存到本地供 pop 页面使用
/// 返回值为一个 map, key 为语言代码，value 为语言名称
async function getGoogleTranslateApiSupportedLanguages(locale) {
  const url = `https://translate.googleapis.com/translate_a/l?client=webapp&hl=${locale}`;
  // Access to fetch at 'https://translate.googleapis.com/translate_a/l?client=webapp&hl=zh-CN' from origin 'chrome-extension://gdignmkpgdklaldmdgaonahopedjbkei' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
  const response = await fetch(url, {
    mode: 'cors'
  });
  const data = await response.json();
  const support_languages = data['tl'];
  // 保存到本地
  chrome.storage.sync
    .set({ support_languages: support_languages })
    .then(() => {});

  return support_languages;
}

function showToast(message) {
  const toast = document.getElementById('toast');
  toast.innerText = message;
  toast.classList.add('show');

  setTimeout(function () {
    toast.classList.remove('show');
  }, 2000);
}


subtitle.dual/scripts/content.js
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

let dual_subs_enabled = false;
let menu_option = null;
let dual_subtitle_track = null;

// 发送消息给 background 的渠道
const port = chrome.runtime.connect();
port.postMessage({ message: 'Nothing' });

// 自动读取用户设置，用于主动触发 双字幕功能
chrome.storage.sync.get('enabled', ({ enabled }) => {
  dual_subs_enabled = enabled;
});

// 被动触发 双字幕功能. by 监听来自 popup 的消息，
chrome.runtime.onMessage.addListener(({ enabled: isEnabled }) => {
  dual_subs_enabled = isEnabled;
});

// 主动触发 双字幕功能，by 监听页面加载完毕事件
window.onload = async () => {
  console.log('window.onload');
  // 因为页面加载和视频加载等异步问题，包括字幕开关被覆盖等问题，逐个解决问题比较麻烦
  // 开一个全局定时器来解决
  setInterval(() => {
    updateSubtitles();
  }, 500);
};

async function updateSubtitles() {
  if (!dual_subs_enabled) {
    display_menu_option(false, menu_option);
    hide_subtitle(dual_subtitle_track);
    return;
  } else {
    display_menu_option(true, menu_option);
    show_subtitle(dual_subtitle_track);
  }

  const videoElement = document.querySelector('video');
  if (!videoElement) {
    return;
  }
  const source_lang = 'en';
  const source_subtitle = get_subtitle_by_lang(videoElement, source_lang);

  const target_lang = 'zh-CN';
  const target_subtitle = get_subtitle_by_lang(videoElement, target_lang);

  const dual_lang = source_lang + '.' + target_lang;
  dual_subtitle_track = get_subtitle_by_lang(videoElement, dual_lang);

  if (source_subtitle && target_subtitle) {
    if (!dual_subtitle_track) {
      dual_subtitle_track = clone_track_to_video(
        videoElement,
        source_subtitle,
        'Dual Subtitle',
        dual_lang
      );
      menu_option = await append_option_to_menu();
    }
  } else if (target_subtitle && !source_subtitle) {
    if (!dual_subtitle_track) {
      dual_subtitle_track = clone_track_to_video(
        videoElement,
        target_subtitle,
        'Dual Subtitle',
        dual_lang
      );
      menu_option = await append_option_to_menu();
    }
  } else if (source_subtitle && !target_subtitle) {
    if (!dual_subtitle_track) {
      dual_subtitle_track = clone_track_to_video(
        videoElement,
        source_subtitle,
        'Dual Subtitle',
        dual_lang
      );
      await sleep(1000);
      set_translated_subtitle(dual_subtitle_track, source_lang, target_lang);
      menu_option = await append_option_to_menu();
    }
  }
}

function hide_subtitle(track) {
  // 调整为 hidden 是为了让字幕不显示在视频上，但是仍然可以通过 track.cues 来获取字幕内容
  if (track && track.mode !== 'hidden') {
    track.mode = 'hidden';
  }
}

function show_subtitle(track) {
  if (track && track.mode !== 'showing') {
    track.mode = 'showing';
  }
}

function clone_track_to_video(videoElement, baseTrack, label, srclang) {
  const clonedTrack = document.createElement('track');
  clonedTrack.kind = baseTrack.kind;
  clonedTrack.label = label;
  clonedTrack.srclang = srclang;
  clonedTrack.track.mode = 'showing';
  // filter the same srclang with base track
  const trackNodes = document.querySelectorAll('track');
  for (const node of trackNodes) {
    if (node.srclang === baseTrack.language) {
      clonedTrack.src = node.src;
      break;
    }
  }
  videoElement.appendChild(clonedTrack);
  return clonedTrack;
}

async function append_option_to_menu() {
  await sleep(1000);
  const subtitleMenu_ul = document.getElementById('subtitle-menu');
  const clonedLi = subtitleMenu_ul.children[0].cloneNode(true);
  subtitleMenu_ul.insertBefore(clonedLi, subtitleMenu_ul.children[0]);
  return clonedLi;
}

function display_menu_option(enabled, li_ele) {
  if (!li_ele) {
    return;
  }
  li_ele.children[0].children[1].setAttribute(
    'aria-label',
    `Dual(${enabled ? 'enabled' : 'disabled'})`
  );
  li_ele.children[0].children[1].textContent = `Dual(${
    enabled ? 'enabled' : 'disabled'
  })`;
}

function get_subtitle_by_lang(videoElement, language) {
  // https://developer.mozilla.org/zh-CN/docs/Web/HTML/Element/track
  const textTracks = videoElement.textTracks;

  if (!textTracks || textTracks.length === 0) {
    return null;
  }

  for (const track of textTracks) {
    if (track.language === language) {
      return track;
    }
  }

  return null;
}

function set_translated_subtitle(track, source_lang, target_lang) {
  let cues = track.cues;
  // 由于逐句翻译会大量请求翻译 API，需要减少请求次数
  const cuesTextList = getCuesTextList(cues);
  // 逐个翻译，并替换原来的字幕
  for (const element of cuesTextList) {
    getTranslation(element[1], source_lang, target_lang, (translatedText) => {
      // 用两个换行符来分割，因为有的视频字幕是自带换行符
      const translatedTextList = translatedText.split('\n\n');
      for (let j = 0; j < translatedTextList.length; j++) {
        cues[element[0] + j].text += '\n' + translatedTextList[j];
      }
    });
  }
}

function getCuesTextList(cues) {
  // 取出字幕的所有文本内容，整合成为一个列表
  // 每项为不大于 5000 字的字符串，（好像目前使用的这个 API 有 5000 字上限？）
  // 以及它在 cues 的起始位置
  // 返回的数据结构大概是 [[0, 文本], [95, 文本]]
  let cuesTextList = [];
  for (let i = 0; i < cues.length; i++) {
    if (
      cuesTextList.length &&
      cuesTextList[cuesTextList.length - 1][1].length + cues[i].text.length <
        4500
    ) {
      // 需要插入一个分隔符(换行)，以便之后为翻译完的字符串 split
      // 用两个换行符来分割，因为有的视频字幕是自带换行符
      cuesTextList[cuesTextList.length - 1][1] += '\n\n' + cues[i].text;
    } else {
      cuesTextList.push([i, cues[i].text]);
    }
  }
  return cuesTextList;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function getTranslation(words, source_lang, target_lang, callback) {
  const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=${source_lang}&tl=${target_lang}&dt=t&q=${encodeURI(
    words
  )}`;
  fetch(url)
    .then((response) => response.json())
    .then((data) => {
      const translatedList = data[0];
      let translatedText = '';
      for (const element of translatedList) {
        translatedText += element[0];
      }
      callback(translatedText);
    })
    .catch((error) => console.error(error));
}


subtitle.dual/scripts/content_apple_dev.js
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

let dual_subs_enabled = false;
let target_subs_has_set = false;
let last_cues_length = 0; // 用于记录上一次字幕长度，用于判断是否有新字幕更新或者切换字幕
let origin_text_track = null;
let origin_cue_ids = [];
const base_sub_lang = 'en';
let target_sub_lang = '';

// 自动读取用户设置，用于主动触发 双字幕功能
chrome.storage.sync.get('enabled', ({ enabled }) => {
  dual_subs_enabled = enabled;
});

chrome.storage.sync.get('target_language', ({ target_language }) => {
  target_sub_lang = target_language;
});

// 被动触发 双字幕功能. by 监听来自 popup 的消息，
chrome.runtime.onMessage.addListener(({ enabled: isEnabled }) => {
  dual_subs_enabled = isEnabled;
});

// 主动触发 双字幕功能，by 监听页面加载完毕事件
window.onload = async () => {
  console.log('window.onload');
  // 因为页面加载和视频加载等异步问题，包括字幕开关被覆盖等问题，逐个解决问题比较麻烦
  // 开一个全局定时器来解决
  setInterval(() => {
    updateSubtitles();
  }, 1000);
};

// WWDC 有目标语言字幕，则直接展示目标语言字幕
// WWDC 没有目标语言字幕，则展示原字幕，并翻译成目标语言，拼接附着在原字幕上
async function updateSubtitles() {
  // 如果没有开启双字幕功能，则不执行
  if (!dual_subs_enabled) return;
  // 如果已经设置过目标语言字幕，则不再执行（因为有目标语言字幕，且设置过一次，后面跟随用户操作即可）
  if (target_subs_has_set) return;

  const htmlVideoElement = document.querySelector('video');
  if (!htmlVideoElement) {
    console.log('htmlVideoElement not found');
    return;
  }

  // 有目标语言字幕，则直接展示目标语言字幕
  let matched_target_sub_lang = target_sub_lang;
  let target_text_track = get_text_track_by_lang(
    htmlVideoElement,
    matched_target_sub_lang
  );
  if (!target_text_track) {
    // 处理中文简体和繁体的问题，Apple Developer 视频只有简体，没有繁体
    // 当 WWDC 有简体zh时，目标语言为 zh-CN zh-TW 时，都展示 zh
    // 当 WWDC 没有简体zh时，按照目标语言分别翻译为 zh-CN 或 zh-TW
    matched_target_sub_lang = matched_target_sub_lang.split('-')[0];
    target_text_track = get_text_track_by_lang(
      htmlVideoElement,
      matched_target_sub_lang
    );
  }

  if (target_text_track) {
    console.log(`target_text_track find and set to ${matched_target_sub_lang}`);
    highlight_text_track_by_lang(htmlVideoElement, matched_target_sub_lang);
    target_subs_has_set = true;
    return;
  }

  // 没有目标语言字幕，则展示原字幕，并翻译成目标语言，拼接附着在原字幕上
  const base_text_track = get_text_track_by_lang(
    htmlVideoElement,
    base_sub_lang
  );

  if (!base_text_track || !base_text_track.cues) {
    console.log('base_text_track not found');
    return;
  }

  const cues = base_text_track.cues;
  if (cues.length === last_cues_length || cues.length === 0) {
    return;
  }

  if (!origin_text_track) {
    // create a text track
    // 用于存放原文用的 track
    origin_text_track = htmlVideoElement.addTextTrack(
      'subtitles',
      '🚫 Backup',
      base_sub_lang
    );
    hide_text_track(origin_text_track);
  }

  // 遍历新增的 cue，如果已经翻译过，则不再添入，因为会重复翻译
  for (const cue of cues) {
    if (origin_cue_ids.includes(cue.id)) {
      continue;
    }

    const origin_cue = new VTTCue(cue.startTime, cue.endTime, cue.text);
    origin_cue.id = cue.id;
    origin_text_track.addCue(origin_cue);
    origin_cue_ids.push(cue.id);
  }

  // 每次都从头翻译
  set_translated_subtitle(
    base_text_track,
    origin_text_track,
    base_sub_lang,
    target_sub_lang
  );

  // 将翻译后的
  last_cues_length = cues.length;
}

function hide_text_track(track) {
  if (track && track.mode !== 'hidden') {
    track.mode = 'hidden';
  }
}

function disable_text_track(track) {
  if (track && track.mode !== 'disabled') {
    track.mode = 'disabled';
  }
}

function show_text_track(track) {
  if (track && track.mode !== 'showing') {
    track.mode = 'showing';
  }
}

function get_text_track_by_lang(videoElement, language) {
  // https://developer.mozilla.org/zh-CN/docs/Web/HTML/Element/track
  const textTracks = videoElement.textTracks;
  console.log('language', language);

  if (!textTracks || textTracks.length === 0) {
    return null;
  }

  for (const track of textTracks) {
    if (track.language === language) {
      return track;
    }
  }

  return null;
}

function highlight_text_track_by_lang(videoElement, language) {
  const textTracks = videoElement.textTracks;

  if (!textTracks || textTracks.length === 0) {
    return;
  }

  for (const track of textTracks) {
    if (track.language === language) {
      show_text_track(track);
    } else {
      disable_text_track(track);
    }
  }
}

function set_translated_subtitle(
  target_track,
  text_track,
  source_lang,
  target_lang
) {
  const text_cues = Array.from(text_track.cues);
  const target_cues = Array.from(target_track.cues);

  let startIndex = text_cues.findIndex((cue) => {
    return cue.translated !== true;
  });
  let endIndex = text_cues.findLastIndex((cue) => {
    return cue.translated !== true;
  });
  startIndex = Math.max(startIndex - 10, 0);
  endIndex = Math.min(endIndex + 10, text_cues.length - 1);

  const need_translated_text_cues = text_cues.slice(startIndex, endIndex + 1);

  const need_translated_target_cues = target_cues.slice(
    startIndex,
    endIndex + 1
  );

  // 由于逐句翻译会大量请求翻译 API，需要减少请求次数
  const cuesTextList = getCuesTextList(need_translated_text_cues);
  // 逐个翻译，并替换原来的字幕
  for (const element of cuesTextList) {
    getTranslation(element[1], source_lang, target_lang, (translatedText) => {
      // 用两个换行符来分割，因为有的视频字幕是自带换行符
      const translatedTextList = translatedText.split('\n\n');
      for (let j = 0; j < translatedTextList.length; j++) {
        need_translated_target_cues[element[0] + j].text =
          // 原文 + 换行 + 译文
          need_translated_text_cues[element[0] + j].text +
          '\n' +
          translatedTextList[j];
        need_translated_text_cues[element[0] + j].translated = true; // 标记已翻译
      }
    });
  }
}

function getCuesTextList(cues) {
  // 取出字幕的所有文本内容，整合成为一个列表
  // 每项为不大于 5000 字的字符串，（好像目前使用的这个 API 有 5000 字上限？）
  // 以及它在 cues 的起始位置
  // 返回的数据结构大概是 [[0, 文本], [95, 文本]]
  let cuesTextList = [];
  for (let i = 0; i < cues.length; i++) {
    if (
      cuesTextList.length &&
      cuesTextList[cuesTextList.length - 1][1].length + cues[i].text.length <
        4500
    ) {
      // 需要插入一个分隔符(换行)，以便之后为翻译完的字符串 split
      // 移除字幕中的换行，因为影响翻译效果，效果增强很多
      cuesTextList[cuesTextList.length - 1][1] +=
        '\n\n' + cues[i].text.replace('\n', '');
    } else {
      cuesTextList.push([i, cues[i].text]);
    }
  }
  return cuesTextList;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function getTranslation(words, source_lang, target_lang, callback) {
  const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=${source_lang}&tl=${target_lang}&dt=t&q=${encodeURI(
    words
  )}`;
  fetch(url)
    .then((response) => response.json())
    .then((data) => {
      const translatedList = data[0];
      let translatedText = '';
      for (const element of translatedList) {
        translatedText += element[0];
      }
      callback(translatedText);
    })
    .catch((error) => console.error(error));
}
